src/documents/posts/2015-04-02-delivery-engineering.md

```diff
@@ -0,0 +1,47 @@
+---
+title: Delivery Engineering
+author: Bob Zoller
+layout: post
+disqus:
+  shortname: goodeggsbytes
+  url: "http://bites.goodeggs.com/post/delivery-engineering"
+---
+
+Delivery Engineering is a term I first heard in the January 2015 [Thoughtworks Radar][thoughtworks-deleng]:
+
+> where there *is* a need for significant investment in [DevOps] tooling and automation, we do see a role for a Delivery Engineering team.  Rather than being a helpdesk, these teams build tooling and enable teams to deploy, monitor, and maintain their own production environments.
+
+In other words, Delivery Engineering enables the _culture_ of DevOps to thrive on each team, rather than silo that knowledge up in one team or role.
+
+I was excited when I read this because it concisely captured the goals of a team I had recently formed, which we were calling (quite opaquely), "Infrastructure."  We formed the Infrastructure team because we felt like we had crossed a threshold as an engineering organization, where it made sense to carve out a set of folks that could own shared tools and services used by all our engineering teams.  Think continuous integration pipelines, node modules for reporting stats, best practices for alerting.
+
+So what's working well and what isn't?
+
+<!-- more -->
+
+Having one team who is accountable for shared services has been an obvious win.  This includes both internal shared services (like a custom deploy server we call Ecru), and thirdparty ones too (like [Travis][travis], [MongoLab][mongolab], [SumoLogic][sumologic], etc).  That's not to say we're the gatekeeper -- no, credentials are still shared and teams often reach out for support directly to those companies.  But our team is there to support our peers, manage upgrades, publish best practices, etc.
+
+Extracting application scaffolding, runtime modules, and build and deploy tools into shared repos has also been valuable.  As much as possible, we try to wrap up these often low-level chunks of code and expose higher-level APIs to our peers.  In the process we share reliability improvements, can swap out underlying services, etc.  In addition to code, we publish best-practices and guides, like "Delivering Software at Good Eggs," which lays out all the steps to putting an app in production.
+
+When we started the Delivery Engineering team, we talked about our values or strategies we'd use.  We decided on three: automation, transparency, and self-service.  Along with those core values, our internal culture of open source has been critical to our success.  We rely heavily on Github Issues, Pull Requests, and Travis' ability to do branch builds and deploys (automate all the things! -- including publishing new versions of internal node modules to a private npm repo).
```

Comment:
I agree, this is pretty compelling. I feel like talking about what our goals and values are for the team would be a great start, and then after the fold, use the ThoughtWorks definition. I think that telling the story about how we had the team before we had a name is interesting to tell.

We "invented" this team at Good Eggs, we didn't get the idea from ThoughtWorks and starting with the ThoughtWorks quote kind of takes away from that.